src/wp-includes/class-wp-block-type.php

```diff
@@ -540,4 +563,30 @@ public function get_attributes() {
 			$this->attributes :
 			array();
 	}
+
+	/**
+	 * Get block variations.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @return array[]
+	 */
+	public function get_variations() {
+		if ( ! isset( $this->variations ) ) {
+			$this->variations = array();
+			if ( is_callable( $this->variation_callback ) ) {
+				$this->variations = call_user_func( $this->variation_callback );
+			}
+		}
+
+		/**
+		 * Filters the registered variations for a block type.
+		 *
+		 * @since 6.5.0
+		 *
+		 * @param array         $variations Array of registered variations for a block type.
+		 * @param WP_Block_Type $block_type The full block type object.
+		 */
+		return apply_filters( 'get_block_type_variations', $this->variations, $this );
```

Comment:
I  checked the latest changes in the PR and everything looks great. It’s quite interesting how you managed to go with fully backward compatible approach that works out of the box with things like REST API endpoint.

My only question is related to this new filter as I couldn’t find information what role it plays. What are the use cases it would cover?

Reply:
@gziolo The main reason to have a filter here is to provide the ability to re-run the callback in case there were any possible changes to the content after the first build of variation.

Reply:
For example, in  - https://github.com/WordPress/gutenberg/pull/54801, the variation is available after the registration. 

Even though what we are trying to solve is not similar to the issue listed in PR#54801, it did make us think about what if there was a change in the variation that was not present during the call to callback.

Reply:
Interesting, I now can see how that might become useful.

Coincidentally, today I commented on the proposal for building a formal API for registering block variations with PHP: https://github.com/WordPress/gutenberg/issues/47170. Do you have any thoughts on how it all could work together with the changes proposed in this PR. It seems like using the new filter would cover some of the use cases.

Reply:
Gave a quick look into https://github.com/WordPress/gutenberg/issues/47170. I believe this filter will solve his problem.  The developer can add a function to run the query looks and modify the variations returned by the API. With this hook, the person could append to the existing variations or modify them completely. 

There is yet another approach if the developer wants to replace variations returned by the core completely (with the help of this PR). The person could change the callback to the function that runs the query and then set the `$variations` to null, just before the API call. This way `get_variations` would execute the custom callback.


Filters plays a role majorly when we want to retain some part of the original variation if not replacing the callback is more performant.

Reply:
Thanks for sharing the issue for a generic PHP helper for registering variations on a block, @gziolo. While I do think that being able to filter the list of registered variations on a block makes this easier, what is being requested in https://github.com/WordPress/gutenberg/issues/47170 is essentially just a utility function that would allow you to add a new variation object to the array stored in `WP_Block_Type::variations`, since currently the variations property essentially acts as a registry.

We could add a new `register_variation` method to this object that could be used to append a new variation to the list. The trick would be that we would need to ensure that any `variation_callback` declared when the block was registered gets run prior to adding any new variations, or temporarily store newly registered variations in a separate property so their presence doesn't keep the callback from running the first time the variations property is accessed nor cause the overhead from those callbacks to be called unnecessarily. Of the two options, I'd prefer the latter.

Reply:
Introducing a new helper method is definitely a separate issue, but I appreciate your feedback on potential ways to implement it taking performance implications into consideration 👍

For what it's worth a similar helper for registering block styles prints instead JS code that moves the processing to the client 😀